Not sure what is wrong with the actual tests in travis, but you can fix flake8
```
auditlog/tests/test_auditlog.py:239:14: E225 missing whitespace around operator
auditlog/tests/test_auditlog.py:275:14: E225 missing whitespace around operator
```

--- a/auditlog/README.rst
+++ b/auditlog/README.rst
@@ -61,6 +61,7 @@ Contributors
 
 * Sebastien Alix <sebastien.alix@osiell.com>
 * Holger Brunn <hbrunn@therp.nl>
+* Pieter Paulussen <pieter.paulussen@me.com>
 
 Images
 ------

--- a/auditlog/__openerp__.py
+++ b/auditlog/__openerp__.py
@@ -4,7 +4,7 @@
 
 {
     'name': "Audit Log",
-    'version': "8.0.1.3.0",
+    'version': "8.0.1.4.0",
     'author': "ABF OSIELL,Odoo Community Association (OCA)",
     'license': "AGPL-3",
     'website': "http://www.osiell.com",

--- a/auditlog/migrations/8.0.1.4.0/pre-migration.py
+++ b/auditlog/migrations/8.0.1.4.0/pre-migration.py
@@ -1,0 +1,67 @@
+# -*- coding: utf-8 -*-
+# © 2018 Pieter Paulussen <pieter_paulussen@me.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+import logging
+
+
+def migrate(cr, version):
+    if not version:
+        return
+    logger = logging.getLogger(__name__)
+    logger.info("Creating columns: auditlog_log (model_name, model_model) "
+                "and auditlog_log_line (field_name, field_description).")
+    cr.execute("""
+    ALTER TABLE auditlog_log
+    ADD COLUMN IF NOT EXISTS model_name VARCHAR,
+    ADD COLUMN IF NOT EXISTS model_model VARCHAR;
+    ALTER TABLE auditlog_log_line
+    ADD COLUMN IF NOT EXISTS field_name VARCHAR,
+    ADD COLUMN IF NOT EXISTS field_description VARCHAR;
+    """)
+
+    logger.info("Creating indexes on auditlog_log column 'model_id' and "
+                "auditlog_log_line column 'field_id'.")
+    cr.execute("""
+    DO $BLOCK$
+    BEGIN
+      BEGIN
+        CREATE INDEX auditlog_log_model_id_index
+          ON auditlog_log (model_id);
+        EXCEPTION
+        WHEN duplicate_table
+          THEN RAISE NOTICE
+          'index ''auditlog_log_model_id_index'' on auditlog_log already '
+          'exists, skipping';
+      END;
+      BEGIN
+        CREATE INDEX auditlog_log_line_field_id_index
+          ON auditlog_log_line (field_id);
+        EXCEPTION
+        WHEN duplicate_table
+          THEN RAISE NOTICE
+          'index ''auditlog_log_line_field_id_index'' on auditlog_log_line '
+          'already exists, skipping';
+      END;
+    END;
+    $BLOCK$;
+    """)
+
+    logger.info("Preemptive fill auditlog_log columns: 'model_name' and "
+                "'model_model'.")
+    cr.execute("""
+    UPDATE auditlog_log al
+    SET model_name = im.name, model_model = im.model
+    FROM ir_model im
+    WHERE im.id = al.model_id
+    """)
+
+    logger.info("Preemtive fill of auditlog_log_line columns: 'field_name' and"
+                " 'field_description'.")
+    cr.execute("""
+    UPDATE auditlog_log_line al
+    SET field_name = imf.name, field_description = imf.field_description
+    FROM ir_model_fields imf
+    WHERE imf.id = al.field_id
+    """)
+    logger.info("Successfully updated auditlog_log and auditlog_log_line "
+                "tables.")

--- a/auditlog/models/http_request.py
+++ b/auditlog/models/http_request.py
@@ -13,7 +13,7 @@ class AuditlogHTTPRequest(models.Model):
     _order = "create_date DESC"
     _rec_name = 'display_name'
 
-    display_name = fields.Char(u"Name", compute="_display_name")
+    display_name = fields.Char(u"Name", compute="_compute_display_name")
     name = fields.Char(u"Path")
     root_url = fields.Char(u"Root URL")
     user_id = fields.Many2one(
@@ -25,7 +25,7 @@ class AuditlogHTTPRequest(models.Model):
         'auditlog.log', 'http_request_id', string=u"Logs")
 
     @api.multi
-    def _display_name(self):
+    def _compute_display_name(self):
         for httprequest in self:
             create_date = fields.Datetime.from_string(httprequest.create_date)
             tz_create_date = fields.Datetime.context_timestamp(

--- a/auditlog/models/http_session.py
+++ b/auditlog/models/http_session.py
@@ -12,7 +12,7 @@ class AuditlogtHTTPSession(models.Model):
     _order = "create_date DESC"
     _rec_name = 'display_name'
 
-    display_name = fields.Char(u"Name", compute="_display_name")
+    display_name = fields.Char(u"Name", compute="_compute_display_name")
     name = fields.Char(u"Session ID", index=True)
     user_id = fields.Many2one(
         'res.users', string=u"User", index=True)
@@ -20,7 +20,7 @@ class AuditlogtHTTPSession(models.Model):
         'auditlog.http.request', 'http_session_id', string=u"HTTP Requests")
 
     @api.multi
-    def _display_name(self):
+    def _compute_display_name(self):
         for httpsession in self:
             create_date = fields.Datetime.from_string(httpsession.create_date)
             tz_create_date = fields.Datetime.context_timestamp(

--- a/auditlog/models/log.py
+++ b/auditlog/models/log.py
@@ -1,7 +1,9 @@
 # -*- coding: utf-8 -*-
 # © 2015 ABF OSIELL <http://osiell.com>
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
-from openerp import models, fields
+from openerp import api, models, fields
+from openerp.tools.translate import _
+from openerp.exceptions import Warning as UserError
 
 
 class AuditlogLog(models.Model):
@@ -11,7 +13,9 @@ class AuditlogLog(models.Model):
 
     name = fields.Char("Resource Name", size=64)
     model_id = fields.Many2one(
-        'ir.model', string=u"Model")
+        'ir.model', string=u"Model", index=True, on_delete='set null')
+    model_name = fields.Char(string=u'Model Name')
+    model_model = fields.Char(string=u'Technical Model Name')
     res_id = fields.Integer(u"Resource ID")
     user_id = fields.Many2one(
         'res.users', string=u"User")
@@ -28,19 +32,60 @@ class AuditlogLog(models.Model):
          ],
         string=u"Type")
 
+    @api.model
+    def create(self, vals):
+        """ Insert model_name and model_model field values upon creation. """
+        if 'model_id' in vals and vals['model_id']:
+            model = self.env['ir.model'].browse(vals['model_id'])
+            vals.update({'model_name': model.name,
+                         'model_model': model.model})
+        return super(AuditlogLog, self).create(vals)
+
+    @api.multi
+    def write(self, vals):
+        """ Update model_name and model_model field values to reflect model_id
+        changes. """
+        if 'model_id' in vals and vals['model_id']:
+            model = self.env['ir.model'].browse(vals['model_id'])
+            vals.update({'model_name': model.name,
+                         'model_model': model.model})
+        return super(AuditlogLog, self).write(vals)
+
 
 class AuditlogLogLine(models.Model):
     _name = 'auditlog.log.line'
     _description = "Auditlog - Log details (fields updated)"
 
     field_id = fields.Many2one(
-        'ir.model.fields', ondelete='cascade', string=u"Field", required=True)
+        'ir.model.fields', ondelete='set null', string=u"Field", index=True)
     log_id = fields.Many2one(
         'auditlog.log', string=u"Log", ondelete='cascade', index=True)
     old_value = fields.Text(u"Old Value")
     new_value = fields.Text(u"New Value")
     old_value_text = fields.Text(u"Old value Text")
     new_value_text = fields.Text(u"New value Text")
-    field_name = fields.Char(u"Technical name", related='field_id.name')
-    field_description = fields.Char(
-        u"Description", related='field_id.field_description')
+    field_name = fields.Char(u"Technical name")
+    field_description = fields.Char(u"Description")
+
+    @api.model
+    def create(self, vals):
+        """ Ensure field_id is not empty on creation and store field_name and
+        field_description. """
+        if 'field_id' not in vals or not vals['field_id']:
+            raise UserError(_("No field defined to create line."))
+        field = self.env['ir.model.fields'].browse(vals['field_id'])
+        vals.update({'field_name': field.name,
+                     'field_description': field.field_description})
+        return super(AuditlogLogLine, self).create(vals)
+
+    @api.multi
+    def write(self, vals):
+        """ Ensure field_id is set during write and update field_name and
+        field_description values. """
+        if 'field_id' in vals:
+            if not vals['field_id']:
+                raise UserError(_("The field 'field_id' cannot be empty."))
+            field = self.env['ir.model.fields'].browse(vals['field_id'])
+            vals.update({'field_name': field.name,
+                         'field_description': field.field_description})
+        return super(AuditlogLogLine, self).write(vals)

--- a/auditlog/models/rule.py
+++ b/auditlog/models/rule.py
@@ -5,7 +5,7 @@
 import logging
 from psycopg2 import ProgrammingError
 from openerp import models, fields, api, modules, _, SUPERUSER_ID, sql_db
-from openerp.exceptions import ValidationError
+from openerp.exceptions import ValidationError, Warning as UserError
 
 FIELDS_BLACKLIST = [
     'id', 'create_uid', 'create_date', 'write_uid', 'write_date',
@@ -50,8 +50,12 @@ class AuditlogRule(models.Model):
 
     name = fields.Char(u"Name", size=32, required=True)
     model_id = fields.Many2one(
-        'ir.model', u"Model", required=True,
-        help=u"Select model for which you want to generate log.")
+        'ir.model', u"Model",
+        help=u"Select model for which you want to generate log.",
+        on_delete='set null',
+        index=True)
+    model_name = fields.Char(string=u'Model Name')
+    model_model = fields.Char(string=u'Technical Model Name')
     user_ids = fields.Many2many(
         'res.users',
         'audittail_rules_users',
@@ -200,7 +204,7 @@ class AuditlogRule(models.Model):
         """Restore original ORM methods of models defined in rules."""
         updated = False
         for rule in self:
-            model_model = self.env[rule.model_id.model]
+            model_model = self.env[rule.model_id.model or rule.model_model]
             for method in ['create', 'read', 'write', 'unlink']:
                 if getattr(rule, 'log_%s' % method) and hasattr(
                         getattr(model_model, method), 'origin'):
@@ -218,8 +222,15 @@ class AuditlogRule(models.Model):
 
     # Unable to find a way to declare the `create` method with the new API,
     # errors occurs with the `_register_hook()` BaseModel method.
+    # pylint: disable=old-api7-method-defined
     def create(self, cr, uid, vals, context=None):
         """Update the registry when a new rule is created."""
+        if 'model_id' not in vals or not vals['model_id']:
+            raise UserError(_("No model defined to create line."))
+        model = self.pool('ir.model').browse(
+            cr, uid, vals['model_id'], context=context)
+        vals.update({'model_name': model.name,
+                     'model_model': model.model})
         res_id = super(AuditlogRule, self).create(
             cr, uid, vals, context=context)
         if self._register_hook(cr, [res_id]):
@@ -228,8 +239,17 @@ class AuditlogRule(models.Model):
 
     # Unable to find a way to declare the `write` method with the new API,
     # errors occurs with the `_register_hook()` BaseModel method.
+    # pylint: disable=old-api7-method-defined
     def write(self, cr, uid, ids, vals, context=None):
         """Update the registry when existing rules are updated."""
+        if 'model_id' in vals:
+            if not vals['model_id']:
+                raise UserError(_("Field 'model_id' cannot be empty."))
+            if vals['model_id']:
+                model = self.pool('ir.model').browse(
+                    cr, uid, vals['model_id'], context=context)
+                vals.update({'model_name': model.name,
+                            'model_model': model.model})
         if isinstance(ids, (int, long)):
             ids = [ids]
         super(AuditlogRule, self).write(cr, uid, ids, vals, context=context)
@@ -686,12 +706,12 @@ class AuditlogRule(models.Model):
             act_window = act_window_model.search(
                 [('name', '=', 'View Log'),
                  ('res_model', '=', 'auditlog.log'),
-                 ('src_model', '=', rule.model_id.model)])
+                 ('src_model', '=', rule.model_id.model or rule.model_model)])
             if act_window:
                 value = 'ir.actions.act_window,%s' % act_window.id
                 act_window.unlink()
                 ir_value = ir_values_model.search(
-                    [('model', '=', rule.model_id.model),
+                    [('model', '=', rule.model_id.model or rule.model_model),
                      ('value', '=', value)])
                 if ir_value:
                     ir_value.unlink()

--- a/auditlog/tests/test_auditlog.py
+++ b/auditlog/tests/test_auditlog.py
@@ -2,6 +2,7 @@
 # © 2015 Therp BV <http://therp.nl>
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 from openerp.tests.common import TransactionCase
+from openerp.addons.base.ir.ir_model import MODULE_UNINSTALL_FLAG
 
 
 class TestAuditlog(object):
@@ -181,3 +182,100 @@ class TestMethods(TransactionCase):
         ])
 
         self.assertEqual(len(logs), 1)
+
+
+class TestFieldRemoval(TransactionCase):
+    def setUp(self):
+        super(TestFieldRemoval, self).setUp()
+
+        # Clear all existing logging lines
+        existing_audit_logs = self.env['auditlog.log'].search([])
+        existing_audit_logs.unlink()
+
+        # Store cursor
+        self.commit_org = self.env.cr.commit
+        self.env.cr.commit = lambda *args: None
+
+        # Create a test model to remove
+        self.test_model = self.env['ir.model'].create({
+            'name': 'x_test_model',
+            'model': "x_test.model",
+            'state': 'manual'
+        })
+
+        # Create a test model field to remove
+        self.test_field = self.env['ir.model.fields'].create({
+            'name': 'x_test_field',
+            'field_description': 'x_Test Field',
+            'model_id': self.test_model.id,
+            'ttype': 'char',
+            'state': 'manual'
+        })
+
+        # Setup auditlog rule
+        self.auditlog_rule = self.env['auditlog.rule'].create({
+            'name': 'test.model',
+            'model_id': self.test_model.id,
+            'log_type': 'fast',
+            'log_read': False,
+            'log_create': True,
+            'log_write': True,
+            'log_unlink': False,
+            'log_custom_method': True,
+        })
+
+        self.auditlog_rule.subscribe()
+
+    def tearDown(self):
+        """ Clean up after testcase. """
+        self.auditlog_rule.unsubscribe()
+        self.auditlog_rule.unlink()
+        self.registry.models.pop('x_test.model')
+        self.registry._pure_function_fields.pop('x_test.model')
+
+        # Restore cursor
+        self.env.cr.commit = self.commit_org
+
+        super(TestFieldRemoval, self).tearDownClass()
+
+    def test_01_field_and_model_removal(self):
+        """ Test field and model removal to check auditlog line persistence """
+        # Trigger log creation
+        rec = self.env['x_test.model'].create({'x_test_field': "test value"})
+        rec.write({'x_test_field': 'test value 2'})
+
+        logs = self.env['auditlog.log'].search([
+            ('res_id', '=', rec.id),
+            ('model_id', '=', self.test_model.id)
+        ])
+
+        # Get the field to remove
+        log_lines = logs.mapped('line_ids')
+        self.assertEqual(len(log_lines), 2)
+
+        # Remove the field
+        self.test_field.with_context({MODULE_UNINSTALL_FLAG: True}).unlink()
+
+        # The log line should still exist
+        self.assertTrue(log_lines)
+
+        # The field should not be linked
+        self.assertFalse(log_lines.mapped('field_id'))
+
+        # The field name and description as well
+        self.assertEqual(log_lines[0].field_name, 'x_test_field')
+        self.assertEqual(log_lines[0].field_description, 'x_Test Field')
+
+        # Remove the model
+        self.test_model.with_context({MODULE_UNINSTALL_FLAG: True}).unlink()
+
+        # Assert log values
+        self.assertTrue(logs)
+        self.assertFalse(logs.mapped('model_id'))
+        self.assertEqual(logs[0].model_name, 'x_test_model')
+        self.assertEqual(logs[0].model_model, 'x_test.model')
+
+        # Assert rule values
+        self.assertFalse(self.auditlog_rule.model_id)
+        self.assertEqual(self.auditlog_rule.model_name, 'x_test_model')
+        self.assertEqual(self.auditlog_rule.model_model, 'x_test.model')

--- a/auditlog/views/auditlog_view.xml
+++ b/auditlog/views/auditlog_view.xml
@@ -133,6 +133,8 @@
                             </group>
                             <group colspan="1">
                                 <field name="model_id" readonly="1"/>
+                                <field name="model_name" attrs="{'invisible': [('model_id', '!=', False)]}"/>
+                                <field name="model_model" attrs="{'invisible': [('model_id', '!=', False)]}"/>
                                 <field name="res_id" readonly="1"/>
                                 <field name="name" readonly="1"/>
                             </group>
@@ -146,6 +148,7 @@
                                 <form string="Log - Field updated">
                                     <group>
                                         <field name="field_id" readonly="1"/>
+                                        <field name="field_name" attrs="{'invisible': [('field_id', '!=', False)]}"/>
                                     </group>
                                     <group string="Values" col="4">
                                         <field name="old_value" readonly="1"/>
@@ -156,7 +159,6 @@
                                 </form>
                                 <tree>
                                     <field name="field_description"/>
-                                    <field name="field_name"/>
                                     <!--<field name="old_value"/>-->
                                     <field name="old_value_text"/>
                                     <!--<field name="new_value"/>-->

